Changes in modelix to make MODELIX-630 work

benedekh opened the pull request:
Changes needed in modelix to make [MODELIX-630](https://github.com/modelix/modelix.mps-plugins/pull/97) work.

languitar commented on model-client/src/commonMain/kotlin/org/modelix/model/client2/ReplicatedModel.kt:
```diff
@@ -251,7 +258,7 @@ private class LocalModel(initialVersion: CLVersion, val idGenerator: IIdGenerato
 
 private class RemoteVersion(val client: IModelClientV2, val branchRef: BranchReference) {
-    private var lastKnownRemoteVersion: CLVersion? = null
+    var lastKnownRemoteVersion: CLVersion? = null
```
Would probably be better to pass this in via the constructor?

benedekh replied:
fixed in [6e976c7](https://github.com/modelix/modelix.core/pull/725/commits/6e976c797ed41c074ade6aaf90ef2b07a7a44c68)

languitar commented on model-client/src/commonMain/kotlin/org/modelix/model/client2/ReplicatedModel.kt:
```diff
@@ -32,20 +32,23 @@ class ReplicatedModel(
     }
 
-    suspend fun start(): IBranch {
+    suspend fun start(providedInitialVersion: CLVersion? = null, onBeforeSync: (IBranch) -> Unit = {}): IBranch {
```
* Wouldn't it be more consistent to also provide `providedInitialVersion` via the constructor, similar to the branch reference?
* Why don't we use the same name as in the `RemoteVersion` object: `lastKnownRemoteVersion`?

slisson replied:
Moving `providedInitialVersion` to the constructor also allows initializing `localModel` earlier and then the `onBeforeSync` callback isn't necessary.

benedekh replied:
We need the `onBeforeSync` call, because we would like to subscribe to changes to the IBranch as soon as it is fetched from the server (via the LocalModel). Is there any other way to accomplish this?

benedekh replied:
fixed in [86dd24c](https://github.com/modelix/modelix.core/pull/725/commits/86dd24c1183e2cf0213864f2d56d0d134077568f), except that the `onBeforeSync` has to stay, because we would like to subscribe to changes on the IBranch, before we fetch the latest version from the server.

languitar replied:
@slisson fine with that now?

slisson replied:
I you provide a `lastKnownRemoteVersion` to the constructor then you can initialize `localModel` already in the constructor instead of in `start`. You can then call `getBranch()` and register your listener before calling `start()`, which means the `onBeforeSync` callback isn't necessary anymore.

benedekh replied:
I'll make this change then.

benedekh replied:
fixed in [9b0ea3f](https://github.com/modelix/modelix.core/pull/725/commits/9b0ea3fa558315b5a31846a9057adb2e484365ba)

benedekh replied:
@slisson is the fix in [9b0ea3f](https://github.com/modelix/modelix.core/pull/725/commits/9b0ea3fa558315b5a31846a9057adb2e484365ba) fine for you?

slisson replied:
I changed the commit a bit. Now it can be merged.

languitar reviewed (changes requested):
Sorry, looking on this again after a few days I see some things we should still think about:

* What about tests for the new feature?
* (One of the) commits should probably be of type `feat` as the user of the model client is now able to do something new.
* What exactly happens when `providedInitialVersion` is given? Some kdoc for the new arguments and the consequences they cause if provided would be good.

benedekh commented:
> (One of the) commits should probably be of type feat as the user of the model client is now able to do something new.
> What exactly happens when providedInitialVersion is given? Some kdoc for the new arguments and the consequences they cause if provided would be good.
> Wouldn't it be more consistent to also provide providedInitialVersion via the constructor, similar to the branch reference?
> Why don't we use the same name as in the RemoteVersion object: lastKnownRemoteVersion?

fixed in [86dd24c](https://github.com/modelix/modelix.core/pull/725/commits/86dd24c1183e2cf0213864f2d56d0d134077568f)

Missing open point that I will take care of later today (hopefully):

> What about tests for the new feature?

benedekh commented:
> What about tests for the new feature?

fixed in [0829ca0](https://github.com/modelix/modelix.core/pull/725/commits/0829ca0afec72d8bccc079e931dcaebcfa44338f)

@languitar I've fixed all your requests. Please re-review this PR, when you have time.

languitar commented on model-client/src/commonMain/kotlin/org/modelix/model/client2/ReplicatedModel.kt:
```diff
@@ -23,29 +23,47 @@ import org.modelix.model.operations.OTBranch
 /**
  * Dispose should be called on this, as otherwise a regular polling will go on.
+ *
+ * @property providedScope the CoroutineScope to use for the suspendable tasks
+ * @property lastKnownRemoteVersion the last version on the server from which we want to start the synchronization
+ * @property scope the CoroutineScope to use for the suspendable tasks. Uses providedScope or defaults to the default dispatchers
+ * @property state the current state of the ReplicatedModel
+ * @property localModel thread-safe maintenance of the local version
+ * @property remoteVersion to send and receive versions from the server
+ * @property pollingJob the coroutine job that periodically checks if there is a new remote version and fetches its content
```
> @ property name
> 
> Documents the property of a class which has the specified name. This tag can be used for documenting properties declared in the primary constructor, where putting a doc comment directly before the property definition would be awkward.

https://kotlinlang.org/docs/kotlin-doc.html#property-name

Looks like properties declared in the class body should be documented directly.

benedekh replied:
in [481f83d](https://github.com/modelix/modelix.core/pull/725/commits/481f83da1fef69c3b946773d198703fcae70ce8e), I documented the public API in the end.

languitar commented on model-server/src/test/kotlin/org/modelix/model/server/ReplicatedModelTest.kt:
```diff
@@ -0,0 +1,154 @@
+            assertEquals(2, twoHelloChildren.size)
+        }
+
+        // Step 4: wait a bit so replicated model can fetch the new versions from the server (in case there would be any -- there is none by the way)
+        delay(500)
```
This looks like it has the potential to result in a flaky unit test if we need waiting for synchronizing. Moreover, from the `runTest` docs:

> On JVM and Native, this function behaves similarly to runBlocking, with the difference that the code that it runs will skip delays. This allows to use [delay](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/index.html) in tests without causing them to take more time than necessary. On JS, this function creates a Promise that executes the test body with the delay-skipping behavior.

So do we really need this here?

benedekh replied:
removed in [481f83d](https://github.com/modelix/modelix.core/pull/725/commits/481f83da1fef69c3b946773d198703fcae70ce8e)

languitar commented on model-server/src/test/kotlin/org/modelix/model/server/ReplicatedModelTest.kt:
```diff
@@ -0,0 +1,154 @@
+        val defaultBranchReference = repositoryId.getBranchReference()
+        val oneHelloChildVersion = addHelloChild(initialVersion, client, defaultBranchReference)
+
+        // Step 1.3: add a new child node to get a newer version
+        val twoHelloChildrenVersion = addHelloChild(oneHelloChildVersion, client, defaultBranchReference)
```
Can't we eliminate this second child? One child (+ 1 version) should be enough to test that the latest version is fetched?

benedekh replied:
removed in [481f83d](https://github.com/modelix/modelix.core/pull/725/commits/481f83da1fef69c3b946773d198703fcae70ce8e)

languitar commented on model-server/src/test/kotlin/org/modelix/model/server/ReplicatedModelTest.kt:
```diff
@@ -0,0 +1,154 @@
+            assertEquals(1, oneHelloChild.size)
+        }
+
+        // Step 4: wait a bit so replicated model can fetch the new versions from the server
+        delay(500)
```
Can we get this deterministic without waiting?

benedekh replied:
Good question. Because the point is, the ReplicatedModel must poll the latest version which is running in a Job started on the CoroutineScope.

Maybe with a blocking wait and a timeout. That way we do not wait 500 ms, but shorter if the response arrive quicker. Otherwise we wait untilt timeout.

benedekh commented on model-server/src/test/kotlin/org/modelix/model/server/ReplicatedModelTest.kt:
```diff
@@ -0,0 +1,154 @@
+        val oneHelloChildVersion = addHelloChild(initialVersion, client, defaultBranchReference)
+
+        // Step 1.3: add a new child node to get a newer version
+        val twoHelloChildrenVersion = addHelloChild(oneHelloChildVersion, client, defaultBranchReference)
```
outdated remark, line does not exist anymore.

benedekh commented on model-server/src/test/kotlin/org/modelix/model/server/ReplicatedModelTest.kt:
```diff
@@ -0,0 +1,154 @@
+        val oneHelloChildVersion = addHelloChild(initialVersion, client, defaultBranchReference)
+
+        // Step 1.3: add a new child node to get a newer version
+        val twoHelloChildrenVersion = addHelloChild(oneHelloChildVersion, client, defaultBranchReference)
```
outdated remark, line does not exist anymore.

languitar commented on model-server/src/test/kotlin/org/modelix/model/server/ReplicatedModelTest.kt:
```diff
@@ -0,0 +1,165 @@
+        // Step 2.1: to avoid version was not created by this client exception
+        val initialVersionClone = newClient.loadVersion(repositoryId, initialVersion.getContentHash(), null)
+
+        val scope = CoroutineScope(Dispatchers.Default)
```
Why is this used now?

benedekh replied:
Because there is no direct way to check if ReplicatedModel has fetched the latest version. Indirectly, I have to wait until the "hello" child appears. 

In order to avoid a fixed wait of 500 ms in the coroutine that is running at line 97, I create a CountdownLatch and start a new coroutine that checks if the "hello" child has arrived. If it has, then it opens the latch. The coroutine running at line 97 will be stopped by the Latch and wait until the Latch is opened or 500 ms has passed whichever occurs first.

languitar replied:
ah, and the single-threaded test dispatcher would be blocked already by waiting on the latch?

slisson commented:
:tada: This PR is included in version 8.1.0 :tada:

The release is available on [GitHub release](https://github.com/modelix/modelix.core/releases/tag/8.1.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
